### PR TITLE
NAS-105932 / 21.04 / Add support for cloudflare acme dns challenge

### PR DESCRIPTION
**Pull request by sonicaj**

### CODE REVIEW

**Comment by william-gr** [comment]
Closing PR for now

**Comment by isometry** [comment]
Any plans to resuscitate this PR? I was about to start drafting a duplicate, but it looks like @sonicaj has already done the work.

**Comment by william-gr** [comment]
The ticket associated still exists and will eventually be worked on.

File: src/middlewared/middlewared/plugins/acme_protocol_/cloudflare.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+from certbot.plugins import dns_common
+from CloudFlare.cloudflare import CloudFlare, CloudFlareAPIError
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/acme_protocol_/cloudflare.py:2]: Do we already have that package in debian control?

**Review by william-gr — APPROVED** [review]

File: src/middlewared/middlewared/plugins/acme_protocol_/cloudflare.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        try:
+            cf.zones.dns_records.post(zone_id, data=data)
+        except CloudFlareAPIError as e:
+            code = int(e)
+            hint = None
+
+            if code == 1009:
+                hint = 'Does your API token have "Zone:DNS:Edit" permissions?'
+
+            self.middleware.logger.error('Encountered CloudFlareAPIError adding TXT record: %d %s', e, e)
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/acme_protocol_/cloudflare.py:29]: Did you mean `code, e` ?

File: src/middlewared/middlewared/plugins/acme_protocol_/cloudflare.py (hunk 1 above)
Comment (sonicaj) [reply src/middlewared/middlewared/plugins/acme_protocol_/cloudflare.py:2]: Yup

**Review by william-gr — APPROVED** [review]

**Pull request merged**